### PR TITLE
Extended syntax: copy propagation

### PR DESCRIPTION
**Pull request by Anabra**
See #32

### CODE REVIEW

File: grin/src/Transformations/ExtendedSyntax/Optimising/CopyPropagation.hs [hunk 1]
```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE LambdaCase, TupleSections, ViewPatterns #-}
+module Transformations.ExtendedSyntax.Optimising.CopyPropagation where
+
+import Control.Monad.State
```
Comment (andorp) [inline grin/src/Transformations/ExtendedSyntax/Optimising/CopyPropagation.hs]: Please use Strict version if you don't rely on laziness heavily.

Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Optimising/CopyPropagation.hs]: Is it beneficial to use the strict version instead? Maybe to avoid space leaks?

Comment (andorp) [reply grin/src/Transformations/ExtendedSyntax/Optimising/CopyPropagation.hs]: Yes. We don't want to drag around thunks.

Comment (Anabra) [inline grin/src/Transformations/ExtendedSyntax/Optimising/CopyPropagation.hs]:
```suggestion
import Control.Monad.State.Strict
```

File: grin/src/Transformations/ExtendedSyntax/Util.hs [hunk 2]
```diff
@@ -106,19 +107,19 @@ mapNameUseExp f = \case
 subst :: Ord a => Map a a -> a -> a
 subst env x = Map.findWithDefault x x env
 
--- variable reference substitution (non recursive)
```
Comment (csabahruska) [inline grin/src/Transformations/ExtendedSyntax/Util.hs]: Why did you remove the none recursive comment? It is important to note that the function operates an a single AST constructor.

File: grin/src/Transformations/ExtendedSyntax/Util.hs [hunk 3]
```diff
@@ -106,19 +107,19 @@ mapNameUseExp f = \case
...
+-- substitute all @Names@s in an @Exp@
```
Comment (Anabra) [inline grin/src/Transformations/ExtendedSyntax/Util.hs]:
```suggestion
-- substitute all @Names@s in an @Exp@ (non-recursive)
```

File: grin/src/Transformations/ExtendedSyntax/Util.hs [hunk 4]
```diff
@@ -106,19 +107,19 @@ mapNameUseExp f = \case
...
 substVarRefExp :: Map Name Name -> Exp -> Exp
 substVarRefExp env = mapNameUseExp (subst env)
 
--- val name substitution (non recursive)
+-- substitute all @Names@s in a @Val@
```
Comment (Anabra) [inline grin/src/Transformations/ExtendedSyntax/Util.hs]:
```suggestion
-- substitute all @Names@s in a @Val@ (non-recursive)
```

File: grin/src/Transformations/ExtendedSyntax/Util.hs [hunk 5]
```diff
@@ -106,19 +107,19 @@ mapNameUseExp f = \case
...
 substNamesVal :: Map Name Name -> Val -> Val
 substNamesVal env = mapNamesVal (subst env)
 
--- val name substitution (non recursive)
+-- specialized version of @subst@ to @Val@s
```
Comment (Anabra) [inline grin/src/Transformations/ExtendedSyntax/Util.hs]:
```suggestion
-- specialized version of @subst@ to @Val@s (non-recursive)
```

File: grin/src/Transformations/ExtendedSyntax/Util.hs [hunk 6]
```diff
@@ -106,19 +107,19 @@ mapNameUseExp f = \case
...
 substValsVal :: Map Val Val -> Val -> Val
 substValsVal env = subst env
 
--- val substitution (non recursive)
+-- substitute all @Val@s in an @Exp@
```
Comment (Anabra) [inline grin/src/Transformations/ExtendedSyntax/Util.hs]:
```suggestion
-- substitute all @Val@s in an @Exp@ (non-recursive)
```

File: grin/src/Transformations/ExtendedSyntax/Util.hs (hunk 2 above)
Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Util.hs]: Good point, I left those out.
